smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/TypeScriptWriter.java

```diff
@@ -114,19 +142,54 @@ public TypeScriptWriter addIgnoredDefaultImport(String name, String from, String
      */
     @Deprecated
     public TypeScriptWriter addImport(String name, String as, String from) {
+        boolean isNodePackage = from.startsWith("node:");
+        boolean isNamespacePackage = from.startsWith("@");
+        boolean isRelativeImport = from.startsWith("/") || from.startsWith(".");
+
+        final boolean isPackageImport =
+            isNodePackage
+                || isNamespacePackage
+                || !isRelativeImport;
+
+        if (isPackageImport) {
+            String[] packageNameSegments = from.split("/");
+            String packageName;
+            if (isNodePackage) {
+                packageName = from.substring("node:".length(), from.indexOf('/'));
+            } else if (isNamespacePackage) {
+                packageName = packageNameSegments[0] + "/" + packageNameSegments[1];
+            } else {
+                packageName = packageNameSegments[0];
+            }
+            if (!EXEMPT_DEPENDENCIES.contains(packageName)
+                && !isNodePackage
+                && getDependencies().stream().noneMatch(dep -> dep.getPackageName().equals(packageName))) {
+                throw new CodegenException(
+                    """
+                    The import %s does not correspond to a registered dependency.
+                    TypeScriptWriter::addDependency() is required before ::addImport().
+                    """.formatted(from)
+                );
+            }
+        }
+
         getImportContainer().addImport(name, as, from);
         return this;
     }
 
     /**
      * Imports a type using an alias from a module only if necessary.
+     * Adds the dependency.
      *
      * @param name Type to import.
      * @param as Alias to refer to the type as.
      * @param from PackageContainer to import the type from.
      * @return Returns the writer.
      */
     public TypeScriptWriter addImport(String name, String as, PackageContainer from) {
+        if (from instanceof Dependency) {
+            addDependency((Dependency) from);
+        }
```

Comment:
In the long run, we should move it to taking `Dependency` as the third parameter.